Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            auto* priv = printOperation->priv;
+
+            // FIXME: this should be done in the parent callback
+            auto preparePrintResponse = std::exchange(priv->preparePrintResponse, std::nullopt);
+            RELEASE_ASSERT(preparePrintResponse && preparePrintResponse->token());
+
+            if (error) {
+                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, error->message) });
+                webkitPrintOperationFinished(printOperation.get());
+                return;
+            }
+
+            auto* printSettings = gtk_print_job_get_settings(printJob);
+            GUniquePtr<char> filename(g_filename_from_uri(gtk_print_settings_get(printSettings, GTK_PRINT_SETTINGS_OUTPUT_URI), NULL, NULL));
+            UnixFileDescriptor fd = UnixFileDescriptor { open(filename.get(), O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
+            if (!fd) {
+                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, "Error opening temporary file") });
+                webkitPrintOperationFinished(printOperation.get());
+                return;
+            }
+
+            GRefPtr<GUnixFDList> fdList(g_unix_fd_list_new());
+            int fdIndex = g_unix_fd_list_append(fdList.get(), fd.value(), NULL);
+
+            GVariantBuilder options;
+            g_variant_builder_init(&options, G_VARIANT_TYPE_VARDICT);
+            g_variant_builder_add(&options, "{sv}", "token", g_variant_new_uint32(*preparePrintResponse->token()));
+
+            // FIXME: title
+            GRefPtr<GVariant> arguments(g_variant_new("(ssha{sv})", "", _("Print"), fdIndex, &options));
+
+            g_dbus_proxy_call_with_unix_fd_list(priv->portalProxy.get(), "Print", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, fdList.get(), NULL,
+                [](GObject* proxy, GAsyncResult* result, gpointer userData) {
+                    auto printOperation = adoptGRef(static_cast<WebKitPrintOperation*>(userData));
+                    GUniqueOutPtr<GError> error;
+                    GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(proxy), result, &error.outPtr()));
+                    if (error)
+                        webkitPrintOperationFailed(printOperation.get(), WTFMove(error));
+                    webkitPrintOperationFinished(printOperation.get());
+                }, printOperation.leakRef());
+        }, printOperation.leakRef(), nullptr);
+    });
+}
+
+static WebKitPrintOperationResponse webkitPrintOperationRunPortalDialog(WebKitPrintOperation* printOperation, GtkWindow* parent)
+{
+    auto* priv = printOperation->priv;
+
+    if (!priv->portalProxy) {
+        GUniqueOutPtr<GError> error;
+
+        priv->portalProxy = adoptGRef(g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_NONE, nullptr,
```

Comment:
Why use the sync version? Isn't it easy enough to use the async version here? You'll need to leakRef the `printOperation` and put the rest of the function inside the callback, but you've already shown you know how to do that.

Reply:
Switched to the async version